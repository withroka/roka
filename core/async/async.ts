/**
 * A library for working with asynchronous operations.
 *
 * This package only provides the {@link [pool]} module to limit the
 * concurrency of promises.
 *
 * ```ts
 * import { pool } from "@roka/async/pool";
 * const results = await pool(
 *   [1, 2, 3, 4, 5, 6],
 *   (x) => Promise.resolve(x * 2),
 *   { concurrency: 2 },
 * );
 * ```
 *
 * ## Modules
 *
 *  -  {@link [pool]}: Limit concurrency.
 *
 * @module async
 */
