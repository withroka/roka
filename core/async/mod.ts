/**
 * Helpers for concurrency.
 *
 * Currently, the only function provided is {@link pool}.
 *
 * @example
 * ```ts
 * import { pool } from "@roka/async/pool";
 * import { assertEquals } from "jsr:@std/assert";
 *
 * const results = await pool(
 *   [1, 2, 3],
 *   (x) => Promise.resolve(x),
 *   { concurrency: 2 },
 * );
 * assertEquals(results, [1, 2, 3]);
 * ```
 *
 * @module
 */

export * from "./pool.ts";
