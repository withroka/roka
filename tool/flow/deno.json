{
  "name": "@roka/flow",
  "version": "0.0.0",
  "exports": {
    ".": "./flow.ts"
  },
  "forge": {
    "main": "flow.ts"
  },
  "permissions": {
    "default": {
      "read": true,
      "write": true,
      "run": ["deno", "git"]
    }
  }
}
