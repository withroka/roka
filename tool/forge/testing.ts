/**
 * This module provides utilities to create packages for testing.
 *
 * ```ts
 * import { tempPackage } from "@roka/forge/testing";
 * await using pkg = await tempPackage({ name: "@scope/name" });
 * ```
 *
 * @module testing
 */

import { type Config, type Package, packageInfo } from "@roka/forge/package";
import { join } from "@std/path";

/**
 * Creates a test package under given directory with given configuration.
 *
 * @example
 *
 * ```ts
 * import { testPackage } from "@roka/forge/testing";
 * import { tempDirectory } from "@roka/testing/temp";
 * import { assertEquals } from "@std/assert";
 *
 * await using directory = await tempDirectory();
 * const pkg = await testPackage(directory.path(), {
 *   name: "@scope/name",
 *   version: "1.2.3",
 * });
 *
 * assertEquals(pkg.name, "name");
 * assertEquals(pkg.version, "1.2.3");
 * ```
 */
export async function testPackage(
  directory: string,
  config: Config & { root?: string },
): Promise<Package> {
  await Deno.mkdir(directory, { recursive: true });
  await Deno.writeTextFile(
    join(directory, "deno.json"),
    JSON.stringify(config, undefined, 2),
  );
  const pkg = await packageInfo({ directory });
  if (config.root) pkg.root = config.root;
  return pkg;
}

/**
 * Creates a test package with given configuration.
 *
 * The package will be created under a temporary directory, which will be
 * automatically removed when the package is disposed.
 *
 * @example
 *
 * ```ts
 * import { tempPackage } from "@roka/forge/testing";
 * import { assertEquals } from "@std/assert";
 *
 * await using pkg = await tempPackage({
 *   name: "@scope/name",
 *   version: "1.2.3",
 * });
 *
 * assertEquals(pkg.name, "name");
 * assertEquals(pkg.version, "1.2.3");
 * ```
 */
export async function tempPackage(
  config: Config,
): Promise<AsyncDisposable & Package> {
  const directory = await Deno.makeTempDir();
  const pkg = await testPackage(directory, config);
  return Object.assign(pkg, {
    [Symbol.asyncDispose]: () => Deno.remove(directory, { recursive: true }),
  });
}
