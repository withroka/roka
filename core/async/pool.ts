import { pooledMap as denoPooledMap } from "@std/async";

/**
 * Maps an array of data to an array of results with concurrency control.
 *
 * @example
 * ```ts
 * import { assertEquals } from "jsr:@std/assert";
 *
 * const results = await pool([1, 2, 3], async (n) => n * 2, { concurrency: 2 });
 * assertEquals(results, [2, 4, 6]);
 * ```
 *
 * @template T The type of the input data.
 * @template R The type of the output data.
 * @param array The array of data to map.
 * @param iteratorFn The function to apply to each element of the array.
 * @param options Configuration options.
 * @param options.concurrency The maximum number of concurrent operations.
 * @returns A promise that resolves to an array of results.
 */
export async function pool<T, R>(
  array: Iterable<T> | AsyncIterable<T>,
  iteratorFn: (data: T) => Promise<R>,
  { concurrency = 1 } = {},
): Promise<R[]> {
  return await Array.fromAsync(denoPooledMap(concurrency, array, iteratorFn));
}
