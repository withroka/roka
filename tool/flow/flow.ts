// deno-lint-ignore-file no-console
/**
 * A fast feedback development tool.
 *
 * Roka ‚Äú**flow**‚Äù eliminates chore work in Deno projects by providing fast,
 * comprehensive code quality checks. The tool uses Deno to format, type-check,
 * lint, and test code, including example code in documentation.
 *
 * ## Usage
 *
 * Run **flow** with `deno run -A jsr:@roka/flow`. You can also install with
 * `deno install -gA jsr:@roka/flow` and run with the flow.
 *
 * ```sh
 * flow
 * ```
 *
 * This formats, types-checks, lints, and tests all project files. Individual
 * checks can be run with subcommands.
 *
 * ```sh
 * flow fmt    # format code
 * flow check  # type-check code
 * flow lint   # lint code
 * flow test   # run tests, and generate coverage report
 * ```
 *
 * {@link https://docs.deno.com/go/config/#formatting | Formatting} and
 * {@link https://docs.deno.com/go/config/#linting | linting} are configured
 * in `deno.json`. The
 * {@link https://docs.deno.com/go/doc/#linting | documentation linter} runs
 * when run at the project root with no arguments.
 *
 * Tests use {@link https://docs.deno.com/go/config/#permissions | permissions}
 * from the `deno.json` file. {@link https://jsr.io/@roka/testing/doc/mock | Mocks}
 * and {@link https://jsr.io/@std/testing/doc/snapshot | snapshots} are
 * generated with all permissions enabled when `‚Äîupdate` is passed.
 *
 * ```sh
 * flow test --update
 * ```
 *
 * _**WARNING**: This tool serves the Roka project. Feel free to make
 * contributions if it is useul to you. The API is unstable and may change._
 *
 * @module flow
 */
import { Command } from "@cliffy/command";
import { pool } from "@roka/async/pool";
import { deno, type DenoOptions, type FileResult, type Info } from "@roka/deno";
import { version } from "@roka/forge/version";
import { find } from "@roka/fs/find";
import { git } from "@roka/git";
import { maybe } from "@roka/maybe";
import { assertExists } from "@std/assert/exists";
import {
  moveCursorUp,
  RESTORE_CURSOR,
  SAVE_CURSOR,
} from "@std/cli/unstable-ansi";
import { associateBy, deepMerge, intersect, sumOf } from "@std/collections";
import { bold, dim, green, red, yellow } from "@std/fmt/colors";

const DESCRIPTION = `
  ${bold("üçÉ flow")}

  An assistant tool for Deno projects that formats, type-checks,
  lints, and tests code.
`;

let verbose = false;

/**
 * Run the `flow` CLI tool.
 *
 * @returns The exit code of the command.
 */
export async function flow(): Promise<number> {
  const cmd = new Command()
    .name("flow")
    .version(await version({ release: true, target: true }))
    .meta("deno", Deno.version.deno)
    .meta("v8", Deno.version.v8)
    .meta("typescript", Deno.version.typescript)
    .description(DESCRIPTION)
    .example("flow", "Format code and run all checks.")
    .example("flow --check", "Check for problems without making changes.")
    .example("flow core/", "Format and check files in the core directory.")
    .example("flow check", "Type-check code.")
    .example("flow fmt", "Format code.")
    .example("flow lint", "Lint code.")
    .example("flow test", "Run tests.")
    .usage("<command> [options]")
    .arguments("[paths...:file]")
    .option("--check", "Check for problems only.", { default: false })
    .option("--verbose", "Print additional information.", {
      hidden: true,
      global: true,
      action: () => verbose = true,
    })
    .action(async ({ check }, ...paths) => {
      const fix = !check;
      const found = await files(paths);
      const cmds = deno(options());
      await run(found, [
        (files) => cmds.fmt(files, { ...options(), check }),
        (files) => cmds.check(files),
        ...paths.length === 0
          ? [(files: string[]) => deno(options()).doc(files, { lint: true })]
          : [],
        (files) => cmds.lint(files, { ...options(), fix }),
      ], { prefix: "Checked" });
      await run(found, [
        (files) => cmds.test(files),
      ], { test: true });
    })
    .command("fmt", fmtCommand())
    .command("check", checkCommand())
    .command("lint", lintCommand())
    .command("test", testCommand());
  const { errors } = await maybe(() => cmd.parse());
  for (const error of errors ?? []) {
    console.error(`‚ùå`, error.message);
    if (verbose) console.error(error);
  }
  return errors ? 1 : 0;
}

function fmtCommand() {
  return new Command()
    .description("Format code, including code blocks in JSDoc.")
    .example("flow fmt", "Format all files.")
    .example("flow fmt --check", "Only check if files are formatted.")
    .example("flow fmt **/*.ts", "Format all TypeScript files.")
    .example("flow fmt core/", "Format files in the core directory.")
    .arguments("[paths...:file]")
    .option("--check", "Check if files are formatted.", { default: false })
    .action(async ({ check }, ...paths) => {
      const found = await files(paths);
      await run(found, [
        (files) => deno(options()).fmt(files, { check }),
      ], { prefix: check ? "Checked formatting in" : "Formatted" });
    });
}

function checkCommand() {
  return new Command()
    .description("Type-check code.")
    .example("flow check", "Type-check all files.")
    .example("flow check **/*.ts", "Type-check all TypeScript files.")
    .example("flow check core/", "Type-check files in the core directory.")
    .arguments("[paths...:file]")
    .action(async (_, ...paths) => {
      const found = await files(paths);
      await run(found, [
        (files) => deno(options()).check(files),
      ], { prefix: "Type-checked" });
    });
}

function lintCommand() {
  return new Command()
    .description("Lint code.")
    .example("flow lint", "Lint all files.")
    .example("flow lint --fix", "Fix any fixable linting problems.")
    .example("flow lint **/*.ts", "Lint all TypeScript files.")
    .example("flow lint core/", "Lint files in the core directory.")
    .arguments("[paths...:file]")
    .option("--fix", "Fix any fixable linting errors.", { default: false })
    .action(async ({ fix }, ...paths) => {
      const found = await files(paths);
      await run(found, [
        ...paths.length === 0
          ? [(files: string[]) => deno(options()).doc(files, { lint: true })]
          : [],
        (files) => deno(options()).lint(files, { fix }),
      ], { prefix: "Linted" });
    });
}

function testCommand() {
  return new Command()
    .description("Run tests using Deno's built-in test runner.")
    .example("flow test", "Run tests in all files.")
    .example("flow test --update", "Run tests and update snapshots and mocks.")
    .example("flow test **/*.test.ts", "Run tests in TypeScript test files.")
    .example("flow test core/", "Run tests in the core directory.")
    .option("--update", "Update snapshots and mocks.", { default: false })
    .arguments("[paths...:file]")
    .action(async ({ update }, ...paths) => {
      const found = await files(paths);
      await run(found, [
        (files) => deno(options()).test(files, { update }),
      ], { test: true });
    });
}

function options(): DenoOptions {
  let reported = false;
  function testLine(report: Partial<Info>) {
    assertExists(report.kind === "test");
    assertExists(report.test);
    const pad = "   ".repeat(report.test.length - 1);
    let line = `${pad}${report.test.at(-1)} ...`;
    if (report.success !== undefined) {
      assertExists(report.status);
      assertExists(report.time);
      const color = report.status === "ok"
        ? green
        : report.success
        ? yellow
        : red;
      line += ` ${color(report.status)} ${dim(`(${report.time})`)}`;
    }
    return line;
  }
  const options: DenoOptions = {
    onError({ message }) {
      console.error();
      console.error(message);
      console.error();
    },
    onDebug({ message }) {
      if (!verbose) return;
      if (!reported) console.log();
      reported = true;
      console.debug(message.split("\n").map((l) => `   ${dim(l)}`).join("\n"));
    },
    onPartialInfo(report) {
      if (!Deno.stdout.isTerminal()) return;
      if (report.kind !== "test" || report.test === undefined) return;
      if (!reported) console.log();
      reported = true;
      console.log(
        SAVE_CURSOR + testLine(report) + RESTORE_CURSOR + moveCursorUp(),
      );
    },
    onInfo(report) {
      if (report.kind !== "test" || report.test === undefined) return;
      if (!reported) console.log();
      reported = true;
      (report.success ? console.log : console.warn)(testLine(report));
      reported = true;
    },
  };
  return options;
}

async function files(paths: string[]): Promise<string[]> {
  const explicit = paths.length > 0;
  let found = await Array.fromAsync(find(explicit ? paths : ["."], {
    type: "file",
    ignore: ["**/.git", "**/node_modules", "**/__testdata__"],
  }));
  if (!explicit) {
    const { value: unignored } = await maybe(() =>
      git().ignore.check(found, { matching: false })
    );
    if (unignored !== undefined) found = intersect(found, unignored);
  }
  if (found.length === 0) throw new Error(`No files found: ${paths.join(" ")}`);
  return found;
}

async function run(
  files: string[],
  fns: ((files: string[]) => Promise<FileResult[]>)[],
  options?: { prefix?: string; test?: boolean },
): Promise<void> {
  const results = Object.values(
    (await pool(fns, (fn) => fn(files)))
      .map((r) => associateBy(r, (f) => f.file))
      .reduce((a, b) => deepMerge(a, b)),
  );
  const [output, error] = message(results, options);
  if (output) console.log(`‚úÖ`, output);
  if (error) throw new Error(error);
}

function message(
  results: FileResult[],
  options?: { prefix?: string; test?: boolean },
): [string | undefined, string | undefined] {
  let { prefix, test } = options ?? {};
  const count = (value: number, name: string) =>
    `${value === 0 ? "no" : value} ${name}${value === 1 ? "" : "s"}`;
  const tests = results.flatMap((r) => r.info)
    .filter((i) => i.kind === "test")
    .filter((t) => t.test.length === 1);
  if (test) prefix = `Ran ${count(tests.length, "test")} from`;
  const errorCount = sumOf(results, (r) => r.error.length);
  let message = errorCount === 0
    ? `${prefix} ${count(results.length, "file")}`
    : `${prefix} ${count(results.length, "file")}, found ${
      count(errorCount, "problem")
    }`;
  if (tests.length > 0) {
    console.log();
    const passingTests = tests.filter((t) => t.success === true).length;
    const failingTests = tests.filter((t) => t.success === false).length;
    message = `${message} ${
      dim(`(${passingTests} passed, ` + `${failingTests} failed)`)
    }`;
  }
  return (errorCount === 0) ? [message, undefined] : [undefined, message];
}

if (import.meta.main) Deno.exit(await flow());
