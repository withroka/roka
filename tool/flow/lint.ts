/**
 * This module provides the {@linkcode lint} function for linting project files.
 *
 * This is currently the same as [`deno lint`](https://docs.deno.com/go/lint).
 *
 * ```ts
 * import { lint } from "@roka/flow/lint";
 *
 * (async () => {
 *   try {
 *     await lint(["file1.ts", "file2.ts"]);
 *   } catch (error) {
 *     console.error("Linting failed:", error);
 *   }
 * });
 * ```
 *
 * @module lint
 */

import { pool } from "@roka/async/pool";
import { block } from "./block.ts";

export async function lint(files: string[]): Promise<void> {
  await pool(files, (file) => block(file, { check: ["lint"] }));
  const cmd = new Deno.Command("deno", {
    args: ["lint", "--quiet", "--permit-no-files", ...files],
    stdin: "null",
    stdout: "null",
    stderr: "piped",
  });
  const { success, stderr } = await cmd.output();
  if (!success) {
    const error = new TextDecoder()
      .decode(stderr)
      .trimEnd();
    throw new Error(error);
  }
}
