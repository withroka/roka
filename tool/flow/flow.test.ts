import { pool } from "@roka/async/pool";
import { find } from "@roka/fs/find";
import { tempDirectory } from "@roka/fs/temp";
import { fakeArgs, fakeConsole } from "@roka/testing/fake";
import { basename, dirname, fromFileUrl, join } from "@std/path";
import { assertSnapshot } from "@std/testing/snapshot";
import { flow } from "./flow.ts";

async function run(context: Deno.TestContext) {
  await using directory = await tempDirectory({ chdir: true });
  const dataDirectory = join(
    dirname(fromFileUrl(context.origin)),
    "__testdata__",
  );
  await pool(
    find([dataDirectory], { type: "file" }),
    (path) => Deno.copyFile(path, basename(directory.path(path))),
  );
  using _args = fakeArgs(
    context.name
      .replaceAll("[valid]", "valid.ts")
      .replaceAll("[invalid-code]", "invalid-code.ts")
      .replaceAll("[invalid-comment]", "invalid-comment.ts")
      .replaceAll("[invalid-doc]", "invalid-doc.md")
      .split(" ").slice(1),
  );
  using console = fakeConsole();
  const code = await flow();
  return {
    code,
    // deno-lint-ignore no-console
    output: console
      .output({ stripAnsi: true, stripCss: true, trimEnd: true, wrap: "\n" })
      .replace(/(?<=\n)((?:.*?):\s*)\d+(\.\d+)+(?:.*)?/g, "$1<version>")
      .replaceAll(Deno.cwd(), "<directory>"),
  };
}

async function test(t: Deno.TestContext) {
  await assertSnapshot(t, await run(t));
}

Deno.test("flow --help", test);
Deno.test("flow [valid]", test);
Deno.test("flow --check [valid]", test);
Deno.test("flow fmt [invalid-code]", test);
Deno.test("flow fmt [invalid-comment]", test);
Deno.test("flow fmt [invalid-doc]", test);
Deno.test("flow fmt --check [valid]", test);
Deno.test("flow fmt --check [invalid-code]", test);
Deno.test("flow fmt --check [invalid-comment]", test);
Deno.test("flow fmt --check [invalid-doc]", test);
Deno.test("flow check [valid]", test);
Deno.test("flow check [invalid-code]", test);
Deno.test("flow check [invalid-comment]", test);
Deno.test("flow check [invalid-doc]", test);
Deno.test("flow lint [valid]", test);
Deno.test("flow lint [invalid-code]", test);
Deno.test("flow lint [invalid-comment]", test);
Deno.test("flow lint [invalid-doc]", test);
Deno.test("flow lint --fix [valid]", test);
Deno.test("flow lint --fix [invalid-code]", test);
Deno.test("flow lint --fix [invalid-comment]", test);
Deno.test("flow lint --fix [invalid-doc]", test);
Deno.test("flow test [valid]", test);
Deno.test("flow test [invalid-code]", test);
Deno.test("flow test [invalid-comment]", test);
Deno.test("flow test [invalid-doc]", test);
