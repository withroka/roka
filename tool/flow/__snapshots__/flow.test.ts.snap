export const snapshot = {};

snapshot[`flow --help 1`] = `
{
  code: 1,
  output: "

Usage:   flow <command> [options]
Version: <version>

deno: <version>
v8: <version>
typescript: <version>

Description:

  üçÉ flow

  An assistant tool for Deno projects that formats, type-checks,
  lints, and tests code.

Options:

  -h, --help     - Show this help.
  -V, --version  - Show the version number for this program.
  --check        - Check for problems only.                   (Default: false)

Commands:

  fmt    [paths...]  - Format code, including code blocks in JSDoc.
  check  [paths...]  - Type-check code.
  lint   [paths...]  - Lint code.
  test   [paths...]  - Run tests using Deno's built-in test runner.

Examples:

  flow         Format code and run all checks.
  flow --check Check for problems without making changes.
  flow core/   Format and check files in the core directory.
  flow check   Type-check code.
  flow fmt     Format code.
  flow lint    Lint code.
  flow test    Run tests.

‚ùå Test case attempted to exit with exit code: 0
",
}
`;

snapshot[`flow [valid] 1`] = `
{
  code: 0,
  output: "
‚úÖ Checked 1 file

example test ... ok (0ms)

‚úÖ Ran 1 test from 1 file (1 passed, 0 failed)
",
}
`;

snapshot[`flow --check [valid] 1`] = `
{
  code: 0,
  output: "
‚úÖ Checked 1 file

example test ... ok (0ms)

‚úÖ Ran 1 test from 1 file (1 passed, 0 failed)
",
}
`;

snapshot[`flow fmt [invalid-code] 1`] = `
{
  code: 0,
  output: "
‚úÖ Formatted 1 file
",
}
`;

snapshot[`flow fmt [invalid-comment] 1`] = `
{
  code: 0,
  output: "
‚úÖ Formatted 1 file
",
}
`;

snapshot[`flow fmt [invalid-doc] 1`] = `
{
  code: 0,
  output: "
‚úÖ Formatted 1 file
",
}
`;

snapshot[`flow fmt --check [valid] 1`] = `
{
  code: 0,
  output: "
‚úÖ Checked formatting in 1 file
",
}
`;

snapshot[`flow fmt --check [invalid-code] 1`] = `
{
  code: 1,
  output: '

from <directory>/invalid-code.ts:
 7 | -function example( ): number {
 7 | +function example(): number {
 8 | -   const a: string = window?.location;
 8 | +  const a: string = window?.location;
 9 | -   return "42";
 9 | +  return "42";

‚ùå Checked formatting in 1 file, found 1 problem
',
}
`;

snapshot[`flow fmt --check [invalid-comment] 1`] = `
{
  code: 1,
  output: '

from <directory>/invalid-comment.ts:4:4:
3 | -function example( ): number {
3 | +function example(): number {
4 | -   const a: string = window?.location;
4 | +  const a: string = window?.location;
5 | -   return "42";
5 | +  return "42";

‚ùå Checked formatting in 1 file, found 1 problem
',
}
`;

snapshot[`flow fmt --check [invalid-doc] 1`] = `
{
  code: 1,
  output: '

from <directory>/invalid-doc.md:
 6 | -function example( ): number {
 6 | +function example(): number {
 7 | -   const a: string = window?.location;
 7 | +  const a: string = window?.location;
 8 | -   return "42";
 8 | +  return "42";

‚ùå Checked formatting in 1 file, found 1 problem
',
}
`;

snapshot[`flow check [valid] 1`] = `
{
  code: 0,
  output: "
‚úÖ Type-checked 1 file
",
}
`;

snapshot[`flow check [invalid-code] 1`] = `
{
  code: 1,
  output: \`

TS2367 [ERROR]: This comparison appears to be unintentional because the types '42' and '43' have no overlap.
  if (42 !== 43) throw new Error("fail");
      ~~~~~~~~~
    at file://<directory>/invalid-code.ts:4:7


TS2322 [ERROR]: Type 'Location' is not assignable to type 'string'.
   const a: string = window?.location;
         ^
    at file://<directory>/invalid-code.ts:8:10


TS2322 [ERROR]: Type 'string' is not assignable to type 'number'.
   return "42";
   ~~~~~~
    at file://<directory>/invalid-code.ts:9:4

‚ùå Type-checked 1 file, found 3 problems
\`,
}
`;

snapshot[`flow check [invalid-comment] 1`] = `
{
  code: 1,
  output: \`

TS2367 [ERROR]: This comparison appears to be unintentional because the types '42' and '43' have no overlap.
if (42 !== 43) throw new Error("fail");
    ~~~~~~~~~
    at file://<directory>/invalid-comment.ts:5:8


TS2322 [ERROR]: Type 'Location' is not assignable to type 'string'.
   const a: string = window?.location;
         ^
    at file://<directory>/invalid-comment.ts:8:13


TS2322 [ERROR]: Type 'string' is not assignable to type 'number'.
   return "42";
   ~~~~~~
    at file://<directory>/invalid-comment.ts:9:7

‚ùå Type-checked 1 file, found 3 problems
\`,
}
`;

snapshot[`flow check [invalid-doc] 1`] = `
{
  code: 1,
  output: \`

TS2367 [ERROR]: This comparison appears to be unintentional because the types '42' and '43' have no overlap.
if (42 !== 43) throw new Error("fail");
    ~~~~~~~~~
    at file://<directory>/invalid-doc.md:4:5


TS2322 [ERROR]: Type 'Location' is not assignable to type 'string'.
   const a: string = window?.location;
         ^
    at file://<directory>/invalid-doc.md:7:10


TS2322 [ERROR]: Type 'string' is not assignable to type 'number'.
   return "42";
   ~~~~~~
    at file://<directory>/invalid-doc.md:8:4

‚ùå Type-checked 1 file, found 3 problems
\`,
}
`;

snapshot[`flow lint [valid] 1`] = `
{
  code: 0,
  output: "
‚úÖ Linted 1 file
",
}
`;

snapshot[`flow lint [invalid-code] 1`] = `
{
  code: 1,
  output: '

error[no-constant-condition]: Use of a constant expressions as conditions is not allowed.
 --> <directory>/invalid-code.ts:4:7
  |
4 |   if (42 !== 43) throw new Error("fail");
  |       ^^^^^^^^^
  = hint: Remove the constant expression

  docs: https://docs.deno.com/lint/rules/no-constant-condition


error[no-unused-vars]: \`example\` is never used
 --> <directory>/invalid-code.ts:7:10
  |
7 | function example( ): number {
  |          ^^^^^^^
  = hint: If this is intentional, prefix it with an underscore like \`_example\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-unused-vars]: \`a\` is never used
 --> <directory>/invalid-code.ts:8:10
  |
8 |    const a: string = window?.location;
  |          ^
  = hint: If this is intentional, prefix it with an underscore like \`_a\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-window]: Window is no longer available in Deno
 --> <directory>/invalid-code.ts:8:22
  |
8 |    const a: string = window?.location;
  |                      ^^^^^^
  = hint: Instead, use \`globalThis\`

  docs: https://docs.deno.com/lint/rules/no-window

‚ùå Linted 1 file, found 4 problems
',
}
`;

snapshot[`flow lint [invalid-comment] 1`] = `
{
  code: 1,
  output: '

error[no-constant-condition]: Use of a constant expressions as conditions is not allowed.
 --> <directory>/invalid-comment.ts:5:8
  |
1 | if (42 !== 43) throw new Error("fail");
  |     ^^^^^^^^^
  = hint: Remove the constant expression

  docs: https://docs.deno.com/lint/rules/no-constant-condition


error[no-unused-vars]: \`example\` is never used
 --> <directory>/invalid-comment.ts:7:13
  |
3 | function example( ): number {
  |          ^^^^^^^
  = hint: If this is intentional, prefix it with an underscore like \`_example\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-unused-vars]: \`a\` is never used
 --> <directory>/invalid-comment.ts:8:13
  |
4 |    const a: string = window?.location;
  |          ^
  = hint: If this is intentional, prefix it with an underscore like \`_a\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-window]: Window is no longer available in Deno
 --> <directory>/invalid-comment.ts:8:25
  |
4 |    const a: string = window?.location;
  |                      ^^^^^^
  = hint: Instead, use \`globalThis\`

  docs: https://docs.deno.com/lint/rules/no-window

‚ùå Linted 1 file, found 4 problems
',
}
`;

snapshot[`flow lint [invalid-doc] 1`] = `
{
  code: 1,
  output: '

error[no-constant-condition]: Use of a constant expressions as conditions is not allowed.
 --> <directory>/invalid-doc.md:4:5
  |
1 | if (42 !== 43) throw new Error("fail");
  |     ^^^^^^^^^
  = hint: Remove the constant expression

  docs: https://docs.deno.com/lint/rules/no-constant-condition


error[no-unused-vars]: \`example\` is never used
 --> <directory>/invalid-doc.md:6:10
  |
3 | function example( ): number {
  |          ^^^^^^^
  = hint: If this is intentional, prefix it with an underscore like \`_example\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-unused-vars]: \`a\` is never used
 --> <directory>/invalid-doc.md:7:10
  |
4 |    const a: string = window?.location;
  |          ^
  = hint: If this is intentional, prefix it with an underscore like \`_a\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-window]: Window is no longer available in Deno
 --> <directory>/invalid-doc.md:7:22
  |
4 |    const a: string = window?.location;
  |                      ^^^^^^
  = hint: Instead, use \`globalThis\`

  docs: https://docs.deno.com/lint/rules/no-window

‚ùå Linted 1 file, found 4 problems
',
}
`;

snapshot[`flow lint --fix [valid] 1`] = `
{
  code: 0,
  output: "
‚úÖ Linted 1 file
",
}
`;

snapshot[`flow lint --fix [invalid-code] 1`] = `
{
  code: 1,
  output: '

error[no-constant-condition]: Use of a constant expressions as conditions is not allowed.
 --> <directory>/invalid-code.ts:4:7
  |
4 |   if (42 !== 43) throw new Error("fail");
  |       ^^^^^^^^^
  = hint: Remove the constant expression

  docs: https://docs.deno.com/lint/rules/no-constant-condition


error[no-unused-vars]: \`example\` is never used
 --> <directory>/invalid-code.ts:7:10
  |
7 | function example( ): number {
  |          ^^^^^^^
  = hint: If this is intentional, prefix it with an underscore like \`_example\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-unused-vars]: \`a\` is never used
 --> <directory>/invalid-code.ts:8:10
  |
8 |    const a: string = globalThis?.location;
  |          ^
  = hint: If this is intentional, prefix it with an underscore like \`_a\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars

‚ùå Linted 1 file, found 3 problems
',
}
`;

snapshot[`flow lint --fix [invalid-comment] 1`] = `
{
  code: 1,
  output: '

error[no-constant-condition]: Use of a constant expressions as conditions is not allowed.
 --> <directory>/invalid-comment.ts:5:8
  |
1 | if (42 !== 43) throw new Error("fail");
  |     ^^^^^^^^^
  = hint: Remove the constant expression

  docs: https://docs.deno.com/lint/rules/no-constant-condition


error[no-unused-vars]: \`example\` is never used
 --> <directory>/invalid-comment.ts:7:13
  |
3 | function example( ): number {
  |          ^^^^^^^
  = hint: If this is intentional, prefix it with an underscore like \`_example\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-unused-vars]: \`a\` is never used
 --> <directory>/invalid-comment.ts:8:13
  |
4 |    const a: string = globalThis?.location;
  |          ^
  = hint: If this is intentional, prefix it with an underscore like \`_a\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars

‚ùå Linted 1 file, found 3 problems
',
}
`;

snapshot[`flow lint --fix [invalid-doc] 1`] = `
{
  code: 1,
  output: '

error[no-constant-condition]: Use of a constant expressions as conditions is not allowed.
 --> <directory>/invalid-doc.md:4:5
  |
1 | if (42 !== 43) throw new Error("fail");
  |     ^^^^^^^^^
  = hint: Remove the constant expression

  docs: https://docs.deno.com/lint/rules/no-constant-condition


error[no-unused-vars]: \`example\` is never used
 --> <directory>/invalid-doc.md:6:10
  |
3 | function example( ): number {
  |          ^^^^^^^
  = hint: If this is intentional, prefix it with an underscore like \`_example\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars


error[no-unused-vars]: \`a\` is never used
 --> <directory>/invalid-doc.md:7:10
  |
4 |    const a: string = globalThis?.location;
  |          ^
  = hint: If this is intentional, prefix it with an underscore like \`_a\`

  docs: https://docs.deno.com/lint/rules/no-unused-vars

‚ùå Linted 1 file, found 3 problems
',
}
`;

snapshot[`flow test [valid] 1`] = `
{
  code: 0,
  output: "

example test ... ok (0ms)

‚úÖ Ran 1 test from 1 file (1 passed, 0 failed)
",
}
`;

snapshot[`flow test [invalid-code] 1`] = `
{
  code: 1,
  output: '

example test ... FAILED (0ms)

example test => ./invalid-code.ts:3:6
error: Error: fail
  if (42 !== 43) throw new Error("fail");
                       ^
    at file://<directory>/invalid-code.ts:4:24


‚ùå Ran 1 test from 1 file, found 1 problem (0 passed, 1 failed)
',
}
`;

snapshot[`flow test [invalid-comment] 1`] = `
{
  code: 1,
  output: '

file://<directory>/invalid-comment.ts\$4-12.ts ... FAILED (0ms)

file://<directory>/invalid-comment.ts\$4-12.ts => ./invalid-comment.ts:4:2
error: Error: fail
    if (42 !== 43) throw new Error("fail");
                         ^
    at file://<directory>/invalid-comment.ts:5:25


‚ùå Ran 1 test from 1 file, found 1 problem (0 passed, 1 failed)
',
}
`;

snapshot[`flow test [invalid-doc] 1`] = `
{
  code: 1,
  output: '

file://<directory>/invalid-doc.md\$3-11.ts ... FAILED (0ms)

file://<directory>/invalid-doc.md\$3-11.ts => ./invalid-doc.md:3:2
error: Error: fail
    if (42 !== 43) throw new Error("fail");
                         ^
    at file://<directory>/invalid-doc.md:4:22


‚ùå Ran 1 test from 1 file, found 1 problem (0 passed, 1 failed)
',
}
`;
